src/interrupt/mod.rs

```diff
@@ -0,0 +1,217 @@
+    fn create_group(
+        &self,
+        ty: InterruptSourceType,
+        base: InterruptIndex,
+        count: InterruptIndex,
+    ) -> Result<Arc<dyn InterruptSourceGroup>>;
+
+    /// Destroy an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object created by
+    /// [create_group()](trait.InterruptManager.html#tymethod.create_group).
+    ///
+    /// Assume the caller takes the responsibility to disable all interrupt sources of the group
+    /// before calling destroy_group(). This assumption helps to simplify InterruptSourceGroup
+    /// implementations.
+    fn destroy_group(&self, group: Arc<dyn InterruptSourceGroup>) -> Result<()>;
+}
+
+/// Trait to manage a group of interrupt sources for a device.
+///
+/// A device may support several types of interrupts, and each type of interrupt may contain one or
+/// multiple continuous interrupt sources. For example, a PCI device may concurrently support:
+/// * Legacy Irq: exactly one interrupt source.
+/// * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+/// * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+///
+/// PCI MSI interrupts of a device may not be configured individually, and must configured as a
+/// whole block. So all interrupts of the same type of a device are abstracted as an
+/// [InterruptSourceGroup](trait.InterruptSourceGroup.html) object, instead of abstracting each
+/// interrupt source as a distinct InterruptSource.
+#[allow(clippy::len_without_is_empty)]
+pub trait InterruptSourceGroup: Send + Sync {
+    /// Get type of interrupt sources managed by the group.
+    fn get_type(&self) -> InterruptSourceType;
+
+    /// Get number of interrupt sources managed by the group.
+    fn len(&self) -> InterruptIndex;
+
+    /// Get base of the assigned Interrupt Source Identifiers.
+    fn get_base(&self) -> InterruptIndex;
+
+    /// Get the irqfd to inject interrupts into the guest by using Linux KVM module.
+    fn get_irqfd(&self, _index: InterruptIndex) -> Option<&EventFd> {
+        None
+    }
+
+    /// Get all irqfs of the interrupt group.
+    fn get_irqfds(&self) -> Option<Vec<&EventFd>> {
+        let mut irqfds = Vec::new();
+        for idx in 0..self.len() {
+            if let Some(fd) = self.get_irqfd(idx) {
+                irqfds.push(fd);
+            }
+        }
+
+        if irqfds.len() as u32 == self.len() {
+            Some(irqfds)
+        } else {
+            None
+        }
+    }
+
+    /// Enable the interrupt sources in the group to generate interrupts.
+    fn enable(&self, configs: &[InterruptSourceConfig]) -> Result<()>;
+
+    /// Disable the interrupt sources in the group to generate interrupts.
+    fn disable(&self) -> Result<()>;
+
+    /// Change the configuration to generate interrupts.
+    ///
+    /// # Arguments
+    /// * index: sub-index into the group.
+    /// * config: configuration data for the interrupt source.
+    fn modify(&self, index: InterruptIndex, config: &InterruptSourceConfig) -> Result<()>;
```

Comment:
can't figure out the purpose of this routine. If it's use for updating InterruptSourceConfig, then why can we not change it directly without invoking this routine? From the comments sounds like it will trigger interrupts, then the naming 'modify' is not good here.

Reply:
It's used to update MSI configuration, and there's no way for us to "directly change the configuration". Think about the way to update MSI config:
1) guest write to MSI register
2) the device backend driver caches MSI registers and calls modify() when all data is ready.
3) the underline interrupt manager, which may be KVM or not, update the irq routing info.
The step 3 is interrupt manager dependent.

Reply:
OK, I see. just the name is not easy to understand.